[spirv] Emit OpLine for rich debug info

jaebaek opened the pull request.

ehsannas commented on tools/clang/lib/SPIRV/SpirvEmitter.cpp:
```diff
@@ -4649,22 +4735,25 @@ SpirvInstruction *SpirvEmitter::doUnaryOperator(const UnaryOperator *expr) {
       };
-      incValue = processEachVectorInMatrix(subExpr, originValue, actOnEachVec);
+      incValue = processEachVectorInMatrix(subExpr, originValue, actOnEachVec,
+                                           expr->getLocStart());
```
should this be `subExpr->getLocStart()`? if so, the callee (`processEachVectorInMatrix`) can find this information from `subExpr`, and there will be no need for the additional parameter passed to the function.

ehsannas commented on tools/clang/lib/SPIRV/SpirvEmitter.cpp:
```diff
@@ -5788,8 +5905,10 @@ SpirvEmitter::tryToAssignToMatrixElements(const Expr *lhs,
-      rhsElem = spvBuilder.createCompositeExtract(elemType, rhs, {i});
+    if (accessor.Count > 1) {
+      rhsElem = spvBuilder.createCompositeExtract(elemType, rhs, {i},
+                                                  lhs->getLocStart());
```
should this be `rhs->getLocStart()` ? seems that the code is extracting elements from rhs.

ehsannas commented on tools/clang/lib/SPIRV/SpirvEmitter.cpp:
```diff
@@ -5815,7 +5934,8 @@ SpirvInstruction *SpirvEmitter::processEachVectorInMatrix(
                                           SpirvInstruction *)>
-        actOnEachVector) {
+        actOnEachVector,
+    SourceLocation loc) {
```
this is clang-formatted ?

ehsannas commented on tools/clang/lib/SPIRV/SpirvEmitter.cpp:
```diff
@@ -6545,7 +6678,7 @@ SpirvEmitter::processIntrinsicCallExpr(const CallExpr *callExpr) {
                                            callExpr->getExprLoc());
 
-    retVal = spvBuilder.createLoad(retType, var);
+    retVal = spvBuilder.createLoad(retType, var, callExpr->getExprLoc());
```
can use `srcLoc` defined on line 6458.

ehsannas commented on tools/clang/lib/SPIRV/SpirvEmitter.cpp:
```diff
@@ -6554,7 +6687,7 @@ SpirvEmitter::processIntrinsicCallExpr(const CallExpr *callExpr) {
         declIdMapper.getBuiltinVar(spv::BuiltIn::SubgroupLocalInvocationId,
                                    retType, callExpr->getExprLoc());
-    retVal = spvBuilder.createLoad(retType, var);
+    retVal = spvBuilder.createLoad(retType, var, callExpr->getExprLoc());
```
can use `srcLoc` instead

ehsannas commented on tools/clang/lib/SPIRV/SpirvEmitter.cpp:
```diff
@@ -7611,7 +7756,8 @@ SpirvEmitter::processIntrinsicLdexp(const CallExpr *callExpr) {
   auto *expInstr = doExpr(callExpr->getArg(1));
-  const auto loc = callExpr->getExprLoc();
+  const auto loc = callExpr->getLocStart();
+  const auto arg0Loc = callExpr->getArg(1)->getLocStart();
```
name should be `arg1Loc` ?

ehsannas commented on tools/clang/lib/SPIRV/SpirvEmitter.cpp:
```diff
@@ -9790,9 +9985,8 @@ bool SpirvEmitter::emitEntryFunctionWrapper(const FunctionDecl *decl,
-      spvBuilder.beginFunction(astContext.VoidTy, funcType,
-                               /*SourceLocation*/ {}, decl->getName());
+  entryFunction = spvBuilder.beginFunction(
+      astContext.VoidTy, funcType, decl->getLocStart(), decl->getName());
```
should this be `decl->getBody()->getLocStart()` ?

ehsannas commented on tools/clang/test/CodeGenSPIRV/spirv.debug.opline.hlsl:
```diff
@@ -16,20 +16,21 @@ uint foo(uint val) {
 
 float4 main(uint val : A) : SV_Target {
-  // CHECK:      OpLine [[file]] 23 12
+  // CHECK:      OpLine [[file]] 24 24
```
why have the line numbers and column numbers changed in this file?

greg-lunarg commented:
Among the benefits of this commit is GPU-assisted Validation which takes advantage of rich debug information to give better information on the location and content of the erroneous code.

ehsannas commented:
Happy to hear that @greg-lunarg ! Hopefully this will land soon

ehsannas commented on tools/clang/lib/SPIRV/InitListHandler.cpp:
```diff
@@ -118,25 +121,26 @@ bool InitListHandler::tryToSplitStruct() {
     auto *extract =
-        spvBuilder.createCompositeExtract(field->getType(), init, {i});
-    fields.push_back(extract);
+        spvBuilder.createCompositeExtract(field->getType(), init, {i}, loc);
```
Sorry. You're right, we cannot use `field->getSourceRange().getBegin()` because that will give us the struct field location.
What I really intended to say was: can we use `init->getSourceLocation()` ?

jaebaek replied:
done

ehsannas commented on tools/clang/lib/SPIRV/InitListHandler.cpp:
```diff
@@ -118,25 +121,26 @@ bool InitListHandler::tryToSplitStruct() {
 
   // Create MemberExpr for each field of the struct
-  llvm::SmallVector<SpirvInstruction *, 4> fields;
+  llvm::SmallVector<std::pair<SpirvInstruction *, SourceLocation>, 4> fields;
```
what I mean was: can we use:
```cpp
auto *init = initializers.back();
const auto loc = init->getSourceLocation();
...
auto *extract = spvBuilder.createCompositeExtract(field->getType(), init, {i}, loc);
fields.emplace_back(extract);
```

jaebaek replied:
done

ehsannas commented on tools/clang/lib/SPIRV/InitListHandler.cpp:
```diff
@@ -168,10 +173,10 @@ bool InitListHandler::tryToSplitConstantArray() {
-    auto *extract = spvBuilder.createCompositeExtract(elemType, init, {i});
-    elements.push_back(extract);
+    auto *extract = spvBuilder.createCompositeExtract(elemType, init, {i}, loc);
+    elements.emplace_back(extract, loc);
```
what I meant was: can you do:
```cpp
auto *extract = spvBuilder.createCompositeExtract(elemType, init, {i}, init->getSourceLocation());
```
?

jaebaek replied:
done

jaebaek commented:
It is weird .. github removed my comments :(

In terms of the function decl issue,
```
void foo(int x); // line 1

int main() {
...
}
void foo(int x)
{  // line 7

}
```

I think the existing code is fine, but it would be nice if I added a test.
I added `tools/clang/test/CodeGenSPIRV/spirv.debug.opline.func_decl.hlsl`.
